### ISSUE TITLE
Reconstruction task: Shape mismatch when using decoder with inputs_embeds and labels

### ISSUE THREAD

**Issue by simplzyu** [simulanics/REFRAG#1]
Hi, thanks for your great work!
I encountered a shape mismatch issue during Continual Pretraining (CPT) phase A: Reconstruction curriculum when freezing the decoder.
Code snippet:
for i, ids in enumerate(chunk_ids):
    inp_emb = e[i].unsqueeze(0).unsqueeze(0)  # [1,1,D]
    labels = ids.unsqueeze(0).to(self.device) # [1,64]
    out = self.decoder(inputs_embeds=inp_emb, labels=labels)Shapes:
• inp_emb.shape = [1, 1, 896]
• labels.shape = [1, 64]

Error: the decoder seems to expect the same sequence length for inputs_embeds and labels. However, here the input embedding length is 1 while the labels length is 64.

**Comment by MartrixG** [comment]
I have the same situation. I think maybe the 'input_embeds' needs concat the labels embedding for decoder

**Comment by simulanics** [comment]
I will get the fix uploaded ASAP. Sorry for the delay  in response - family member in critical condition as we prepare for the worst.  Thanks for your patience. 🙏

**Comment by simplzyu** [comment]
Hi, really sorry to hear about your family situation — wishing you strength in this difficult time, and thank you for still taking the time to respond. 🙏 
Also,Thank you for pushing the fix for cpt_recon.
I noticed a similar shape mismatch in cpt_next when computing loss_next_para:
inp.shape = [...]
labels.shape = [...]

**Comment by simulanics** [comment]
> Hi, really sorry to hear about your family situation — wishing you strength in this difficult time, and thank you for still taking the time to respond. 🙏 
> Also,Thank you for pushing the fix for cpt_recon.
> I noticed a similar shape mismatch in cpt_next when computing loss_next_para:
> inp.shape = [...]
> labels.shape = [...]
> 

Thank you 🙏 Truly.

I'll will get this fix issued to the main branch ASAP. I have accounted for the current and most recent few changes to how the HF causal loss is handled. I'll get this fully tested, but the fix is as follows.

Replace the function with this updated version.

```
def loss_next_para(self, full_text: str, s: int, o: int, k: int, expand_frac: float = 0.0) -> torch.Tensor:
    """
    Feed the first s tokens (compressed/expanded context) and predict the next o tokens.
    Fixes: teacher-force by concatenating target token embeddings to inputs and mask labels
    so only the target span contributes to loss.

    Shapes after fix:
      ctx_emb:   [1, T_ctx, D]
      tgt_emb:   [1, T_tgt, D]   (T_tgt == len(out_ids) <= o)
      inputs:    [1, T_ctx + T_tgt, D]
      labels:    [1, T_ctx + T_tgt]   (labels[:, :T_ctx] = -100; labels[:, T_ctx:] = out_ids)
    """
    # Tokenize up to s + o and slice context/target
    toks = self.decoder_tok(full_text, truncation=True, max_length=s + o, return_tensors="pt")
    ids = toks.input_ids[0].to(self.device)

    # If there aren't enough tokens for a context, skip with 0-loss
    if ids.size(0) < max(2, s + 1):
        return torch.tensor(0.0, device=self.device, requires_grad=True)

    ctx_ids = ids[:s]
    out_ids = ids[s:s + o]  # may be shorter than o due to truncation

    # Build (compressed/expanded) context embedding sequence
    ctx_str = self.decoder_tok.decode(ctx_ids, skip_special_tokens=True)
    chunk_strs, chunk_ids = self._chunk_text(ctx_str, k_tokens=k)

    # Encoder→projector path for compressed chunks
    c = self._encode_chunks(chunk_strs)  # [L, D_enc]
    e = self._project_chunks(c)          # [L, D_dec]

    L = len(chunk_ids)
    expand_mask = torch.zeros(L, dtype=torch.bool, device=self.device)
    if L > 0 and expand_frac > 0.0:
        # simple length-based expansion: expand the longest chunks up to expand_frac
        top = max(1, int(round(expand_frac * L)))
        lengths = torch.tensor([len(ch) for ch in chunk_ids], device=self.device)
        top_idx = torch.topk(lengths, k=min(top, L)).indices
        expand_mask[top_idx] = True

    seq = []
    for i, ids_i in enumerate(chunk_ids):
        if expand_mask[i]:
            # expanded: real token embeddings for this chunk
            seq.append(self._decoder_token_embeddings(ids_i.unsqueeze(0)).squeeze(0))  # [t_i, D]
        else:
            # compressed: single slot
            seq.append(e[i].unsqueeze(0))  # [1, D]

    if len(seq) == 0:
        # fallback: if no chunks (short context), embed ctx_ids directly
        seq.append(self._decoder_token_embeddings(ctx_ids.unsqueeze(0)).squeeze(0))  # [s, D]

    ctx_emb = torch.cat(seq, dim=0).unsqueeze(0).contiguous()  # [1, T_ctx, D]
    T_ctx = ctx_emb.size(1)

    # Teacher-forced target embeddings
    if out_ids.numel() == 0:
        # nothing to predict; return 0-loss to keep training loop simple
        return torch.tensor(0.0, device=self.device, requires_grad=True)

    tgt_ids = out_ids.unsqueeze(0).to(self.device)            # [1, T_tgt]
    tgt_emb = self._decoder_token_embeddings(tgt_ids)         # [1, T_tgt, D]
    T_tgt = tgt_emb.size(1)

    # Concatenate context + target embeddings for a single forward
    inputs = torch.cat([ctx_emb, tgt_emb], dim=1)             # [1, T_ctx + T_tgt, D]

    # Attention mask (all ones)
    attn_mask = torch.ones((1, T_ctx + T_tgt), dtype=torch.long, device=self.device)

    # Labels: ignore context positions, supervise only the target span
    labels = torch.full((1, T_ctx + T_tgt), -100, dtype=torch.long, device=self.device)
    labels[0, T_ctx:T_ctx + T_tgt] = out_ids                  # align target to the last T_tgt positions

    out = self.decoder(inputs_embeds=inputs, attention_mask=attn_mask, labels=labels)
    return out.loss
```
